File: 2. 两数相加.cpp

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* h = new ListNode(0);
        ListNode* p = h;
        int tmp = 0;
        while (l1 && l2) {
            p->next = new ListNode((l1->val + l2->val + tmp) % 10);
            tmp = (l1->val + l2->val + tmp) / 10;
            p = p->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1) {
            p->next = new ListNode((l1->val + tmp) %10);
            tmp = (l1->val + tmp) / 10;
            p = p->next;
            l1 = l1->next;
        }
        while (l2) {
            p->next = new ListNode((l2->val + tmp) % 10);
            tmp = (l2->val + tmp) / 10;
            p = p->next;
            l2 = l2->next;
        }
        if (tmp) {
            p->next = new ListNode(1);
        }
        return h->next;
    }
};

File: Golang/2.两数相加.go
package leetcode

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	h := &ListNode{Val: 0}
	p := h
	tmp := 0
	for l1 != nil && l2 != nil {
		p.Next = &ListNode{Val: (l1.Val + l2.Val + tmp) % 10}
		tmp = (l1.Val + l2.Val + tmp) / 10
		p = p.Next
		l1 = l1.Next
		l2 = l2.Next
	}
	for l1 != nil {
		p.Next = &ListNode{Val: (l1.Val + tmp) % 10}
		tmp = (l1.Val + tmp) / 10
		p = p.Next
		l1 = l1.Next
	}
	for l2 != nil {
		p.Next = &ListNode{Val: (l2.Val + tmp) % 10}
		tmp = (l2.Val + tmp) / 10
		p = p.Next
		l2 = l2.Next
	}
	if tmp == 1 {
		p.Next = &ListNode{Val: 1}
	}
	return h.Next
}
